docs/RFCS/20201020_sql_query_denylist.md

```diff
@@ -0,0 +1,183 @@
+
+In order to check incoming queries against the denylist before parsing, more significant changes would need to be made in [the interface between pgwire/conn and connExecutor](https://github.com/cockroachdb/cockroach/blob/47a5c7e89384b5cf7a634e9f77d09f8416f16de1/pkg/sql/pgwire/conn.go#L203) — the parsing step would need to be lifted out of the `pgwire` package into the `connExecutor` so that `pgwire` only sends strings in. Then, `connExecutor` would first perform the regex pattern matching against the denylist; and then afterwards the parsing and execution of commands that are not denied.
+
+The current implementation was chosen because the overhead incurred by an extra parsing step is not significant, and a more invasive change to the `pgwire`/`connExecutor` interface was not warranted. Additionally, this overhead would only be incurred in the case of an actual match of a query to the denylist, which should be quite infrequent (much more infrequent than the case of matching all queries against the denylist as discussed in the “Drawbacks” section above). 
+
+### Using file-managed denylist vs. cluster settings/system table
+A file-managed system for the denylist was chosen over cluster settings and system table for this first iteration of the denylist. This is because the most urgent use case for our users is a situation in which clusters are entirely unresponsive, and therefore updating a denylist via cluster settings or system table would not be feasible. The file-managed denylist will mitigate the most extreme circumstance and therefore is the first denylist management system we will design and implement.
+
+In the future, it is likely that we will add implementations of the denylist via cluster settings for broader use cases when the cluster is not unresponsive. At this point it would be possible to surface the functionality to the Web UI for an easier user experience.
+
+### Denylist implementation in the database vs. SQL proxy
+The concept of a global denylist may make more sense to live in a SQL proxy layer on top of the database, as it is generally not considered part and parcel of core database functionality. Currently we do not support or ship a SQL proxy layer like this on our core database product; we do have one for Cockroach Cloud, but we want to add the functionality to the database because expect that this feature would provide equal value to our on-prem customers. 
+
+The denylist on its own does not seem to warrant the addition of an additional abstraction in the form of a SQL proxy layer to our database product and we consider it out of scope of this RFC. If such a layer were to be added in the future, the denylist functionality should be moved there. 
+
+## Unresolved questions
```

Comment:
could we add a section about testing? would be nice to see what sort of automated tests we can do. and also maybe some manual test cases to verify it works at the end and is a good experience

Reply:
I'll look into this. Thanks for the suggestion 💖